include/nbl/builtin/hlsl/bxdf/ndf/beckmann.hlsl

```diff
@@ -0,0 +1,179 @@
+
+// TODO: get beta from lgamma, see: https://www.cec.uchile.cl/cinetica/pcordero/MC_libros/NumericalRecipesinC.pdf
+
+// TODO: use query_type for D, lambda, beta, DG1 when that's implemented
+template<typename T>
+NBL_PARTIAL_REQ_TOP(concepts::FloatingPointScalar<T>)
+struct Beckmann<T,false NBL_PARTIAL_REQ_BOT(concepts::FloatingPointScalar<T>) >
+{
+    using scalar_type = T;
+    using this_t = Beckmann<T,false>;
+
+    scalar_type D(scalar_type a2, scalar_type NdotH2)
+    {
+        scalar_type nom = exp<scalar_type>( (NdotH2 - scalar_type(1.0)) / (a2 * NdotH2) );   // exp(x) == exp2(x/log(2)) ?
+        scalar_type denom = a2 * NdotH2 * NdotH2;
+        return numbers::inv_pi<scalar_type> * nom / denom;
+    }
+
+    // brdf
+    scalar_type DG1(scalar_type ndf, scalar_type maxNdotV, scalar_type lambda_V)
+    {
+        onePlusLambda_V = scalar_type(1.0) + lambda_V;
+        return ndf::microfacet_to_light_measure_transform<this_t,ndf::MTT_REFLECT>::__call(ndf / onePlusLambda_V, maxNdotV);
+    }
+
+    // bsdf
+    scalar_type DG1(scalar_type ndf, scalar_type absNdotV, scalar_type lambda_V, bool transmitted, scalar_type VdotH, scalar_type LdotH, scalar_type VdotHLdotH, scalar_type orientedEta, scalar_type reflectance)
+    {
+        onePlusLambda_V = scalar_type(1.0) + lambda_V;
+        return ndf::microfacet_to_light_measure_transform<this_t,ndf::MTT_REFLECT_REFRACT>::__call(hlsl::mix(reflectance, scalar_type(1.0) - reflectance, transmitted) * ndf / onePlusLambda_V, absNdotV, transmitted, VdotH, LdotH, VdotHLdotH, orientedEta);
+    }
+
+    scalar_type G1(scalar_type lambda)
+    {
+        return scalar_type(1.0) / (scalar_type(1.0) + lambda);
+    }
+
+    scalar_type C2(scalar_type NdotX2, scalar_type a2)
+    {
+        return NdotX2 / (a2 * (scalar_type(1.0) - NdotX2));
+    }
+
+    scalar_type Lambda(scalar_type c2)
+    {
+        scalar_type c = sqrt<scalar_type>(c2);
+        scalar_type nom = scalar_type(1.0) - scalar_type(1.259) * c + scalar_type(0.396) * c2;
+        scalar_type denom = scalar_type(2.181) * c2 + scalar_type(3.535) * c;
+        return hlsl::mix<scalar_type>(scalar_type(0.0), nom / denom, c < scalar_type(1.6));
+    }
+
+    scalar_type Lambda(scalar_type NdotX2, scalar_type a2)
+    {
+        return Lambda(C2(NdotX2, a2));
+    }
+
+    scalar_type correlated(scalar_type a2, scalar_type NdotV2, scalar_type NdotL2)
+    {
+        scalar_type c2 = C2(NdotV2, a2);
+        scalar_type L_v = Lambda(c2);
+        c2 = C2(NdotL2, a2);
+        scalar_type L_l = Lambda(c2);
+        return G1(L_v + L_l);
+    }
+
+    scalar_type G2_over_G1(scalar_type a2, bool transmitted, scalar_type NdotL2, scalar_type lambdaV_plus_one)
+    {
+        scalar_type lambdaL = Lambda(NdotL2, a2);
+        return hlsl::mix(
+            lambdaV_plus_one / (lambdaV_plus_one + lambdaL),
+            lambdaV_plus_one * hlsl::beta<scalar_type>(lambdaV_plus_one, scalar_type(1.0) + lambdaL),
+            transmitted
+        );
```

Comment:
you should do `lambdaV_plus_one * hlsl::mix(reciprocal(lambda_V_plus_one+lambdaL),beta(),transmitted)`